news/_posts/2020-02-04-the-new-prestashop-test-framework.md

```diff
@@ -0,0 +1,96 @@
+
+There were some concerns with the code too: although it was working, there was no real framework behind it. No separation of concerns, no proper use of inheritance, no distinction between the test logic and the page logic. It was very complicated to find the source of a problem sometimes, and with more than 8000 test cases it frequently was not possible to fix it.
+
+All these reasons made it clear that it was time to drop the whole thing and start over.
+
+
+## What is coming
+
+The QA team decided to work on three points.
+
+### Automation tool
+
+It was already decided to drop Selenium, but what would be a good replacement ? There are a lot of options there, including SaaS solutions (Cypress, QAWolf, etc). After reviewing a few possibilities, the best choice was to go for [Puppeteer](https://github.com/puppeteer/puppeteer).
+
+Puppeteer is a Node library which provides a nice API to control Chrome. It can pretty much do everything you can do manually. It’s maintained by Google, which is always a good sign.
+It’s also more robust and more stable in headless (which was a killer feature for us) since it was designed with headless in mind.
+
+
+### New framework
+
+The aim was also to create a more robust way to create and maintain tests to avoid the problems we had with the last implementation. Since Puppeteer is “just” a library, the QA team would have to create a whole framework around it, and that was a good opportunity to take a look at the [Page Object Pattern](https://martinfowler.com/bliki/PageObject.html). This pattern makes you separate the test logic from the page logic.
+
+On one hand, you create separate objects for every page you’ll need to automate things on. Each object will then hold its own logic, from selectors to basic methods to use on the page (for example, fill a form, sort a table, choose an option from a dropdown, etc).
+
+On the other hand, you create your test logic by using these objects. The goal is to have really small test steps which then use the methods from the objects. This way, your tests don’t know what’s happening behind the scenes. They only call methods which act as an abstraction layer.
+
+This is a good illustration of the separation of concerns: when a page is updated, you only change the corresponding object in its separate file so that selectors and methods still work the same. When your test needs an update, you only change the logic in the steps.
+
+
+### Re-focus the test effort
+
+The third point was about the campaign itself. We wanted to redirect the campaign to be more effective, faster (the full Selenium campaign could last for up to 20 hours !) and easier to maintain. That would mean a shift in the focus of the campaign: instead of making a “do-it-all” campaign, we would split up the effort in 3 campaigns: 
```

Comment:
```suggestion
The third point was about the campaign itself. We wanted to redirect the campaign to be more effective, faster (the full Selenium campaign could last for up to 20 hours!) and easier to maintain. That would mean a shift in the focus of the campaign: instead of making a “do-it-all” campaign, we split up the effort in 3 campaigns: 
```